in_toto/verifylib.py

```diff
@@ -1043,70 +883,84 @@ def verify_disallow_rule(rule, source_artifacts_queue):
     None.
 
   """
-  rule_data = in_toto.rulelib.unpack_rule(rule)
+  filtered_artifacts = fnmatch.filter(artifacts_queue, rule_pattern)
 
-  matched_artifacts = fnmatch.filter(
-      source_artifacts_queue, rule_data["pattern"])
+  if len(filtered_artifacts):
+    raise RuleVerificationError("'DISALLOW {}' matched the following "
+      "artifacts: {}\n{}".format(rule_pattern, filtered_artifacts,
+      _get_artifact_rule_traceback()))
 
-  if len(matched_artifacts):
-    raise RuleVerificationError("Rule '{0}' failed.\nRule pattern matches the"
-        " following artifacts of the artifact queue, which is disallowed:\n"
-        " {1} \n{2}".format(" ".join(rule), matched_artifacts,
-            get_traceback_info()))
 
+def _get_artifact_rule_traceback():
+  """Build and return string form global `RULE_TRACE` which may be used as
+  error message for RuleVerificationError.
 
-def verify_item_rules(source_name, source_type, rules, links):
   """
-  <Purpose>
-    Iteratively apply all passed material or product rules of one item (step or
-    inspection) to enforce and authorize artifacts reported by the
-    corresponding link and/or to guarantee that artifacts are linked together
-    across links.
-    In the beginning all artifacts are placed in a queue according to their
-    type. If an artifact gets consumed by a rule it is removed from the queue,
-    hence an artifact can only be consumed once.
+  traceback_str = "Full trace for 'expected_{0}' of item '{1}':\n".format(
+      RULE_TRACE["source_type"], RULE_TRACE["source_name"])
+
+  # Show all materials and products available in the beginning and
+  # label the one that is used to generate a queue.
+  for source_type in ["materials", "products"]:
+    traceback_str += "Available {}{}:\n{}\n".format(
+        source_type,
+        [" (used for queue)", ""][RULE_TRACE["source_type"] != source_type],
+        RULE_TRACE[source_type])
+
+  for trace_entry in RULE_TRACE["trace"]:
+    traceback_str += "Queue after '{0}':\n".format(
+        " ".join(trace_entry["rule"]))
+    traceback_str += "{}\n".format(trace_entry["queue"])
```

Comment:
Why? To avoid having a global variable? There are only two places that access it -- `verify_item_rules` writes to it, and `_get_artifact_rule_traceback` reads from it. I think a decorator might be a bit of an overkill.

On a side note, we could easily remove the rule queue trace and instead update the logging statements to provide the same information. See https://github.com/in-toto/in-toto/pull/243#issuecomment-444912500 ff.